pkg/operator/certrotation/target_test.go

```diff
@@ -631,3 +636,217 @@ func TestEnsureTargetSignerCertKeyPair(t *testing.T) {
 		}
 	}
 }
+
+// A simplified test scenario for a race condition when multiple CertRotationControllers
+// are used with the same SignerSecret (CA) and CABundleConfigMap.
+//
+// Let's assume there are two target controllers (firstCertController and secondCertController).
+// The specific implementation details of these controllers are not defined in this test.
+//
+// During the recovery process, when the Signer is expired, the following sequence of events can occur:
+//
+// 1. firstCertController regenerates the signer (firstSignerSecret).
+//
+// 2. secondCertController regenerates the signer (secondSignerSecret) because it hasn't observed the change from step 1.
+//
+// 3. firstCertController regenerates the client cert and signs it with the firstSignerSecret.
+//
+// 4. secondCertController regenerates the client cert and signs it with the secondSignerSecret.
+//
+// At this point, we have client certs signed with two different signers (different keys) but with the same CommonName.
+//
+//  5. Since secondCertController persisted its signer last (the secondSignerSecret is stored in the shared secret), at some point,
+//     the firstCertController gets triggered BUT it will NOT regenerate the client cert BECAUSE
+//     the CommonName (from the firstSignerSecret) matches the one from secondSignerSecret.
+func TestEnsureTargetCertKeyPairSignerCommonName(t *testing.T) {
+	// firstSignerSecret would be generated by the firstCertController
+	firstSignerSecret := &corev1.Secret{}
+	require.NoError(t, setSigningCertKeyPairSecret(firstSignerSecret, time.Hour))
+
+	// the secondSignerSecret would be generated by the secondCertController
+	secondSignerSecret := &corev1.Secret{}
+	require.NoError(t, setSigningCertKeyPairSecret(secondSignerSecret, time.Hour))
+
+	firstSecretCommonName := firstSignerSecret.Annotations[CertificateIssuer]
+	secondSecretCommonName := secondSignerSecret.Annotations[CertificateIssuer]
+	require.Equal(t, firstSecretCommonName, secondSecretCommonName)
```

Comment:
This relies on common name to include unix timestamp, so it would pass in the unit test, but not real run

Reply:
Right, what I'm saying is this test is not really indicative for the fix, the actual solution would prevent this situation from happening - so later on there is no way to fix the test

Reply:
I don't intend to merge this PR. The tests here helped me understood where the race condition is as described in the linked docs.

Reply:
> the actual solution would prevent this situation from happening - so later on there is no way to fix the test

this is not true, switching from `apply` to `update` will not change the function that generates the common name.

Reply:
Confusingly we don't do `apply` in `applySecret`, its using `Update` - https://github.com/openshift/library-go/blob/e0aa70d556789b3ff5e9af85e838320b3063aa14/pkg/operator/resource/resourceapply/core.go#L458

We can do update it to use `get with specific revision` but I'm struggling to write a unit test for it

Reply:
We don't need to pass an additional `RV`. The RV is read from the resource. The optimistic lock is built-in. Give it a try.
Also switch the CABundle to use update semantic instead of apply.

Reply:
>We don't need to pass an additional RV. The RV is read from the resource Give it a try.

In that case just using "optimistic lock" is insufficient. 

CABundle uses `applyConfigMap` which is also Update: https://github.com/vrutkovs/library-go/blob/f8b331e2b5bcb147092181769e1094494f663574/pkg/operator/resource/resourceapply/core.go#L348

Reply:
the apply function is more like overwrite, we need to change the controller/struct/type to issue an update for a resource it has currently reconciled.

Reply:
>we need to change the controller/struct/type to issue an update for a resource it has currently reconciled

This is what its doing now, isn't it? I don't quite understand what needs fixing and why the bug resurfaces for this controller only - it sounds like it could have happened across any library-go user but it didn't?

In any case, we have an alternative solution which has unittests and passing e2e tests - https://github.com/openshift/library-go/pull/1722

Reply:
> I don't quite understand what needs fixing and why the bug resurfaces for this controller only

Have a look at the doc linked in the description of this PR. It contains more details about the possible race condition.

> In any case, we have an alternative solution which has unittests and passing e2e tests - https://github.com/openshift/library-go/pull/1722

#1722 is not going to solve the issue because the controllers run across multiple processes. 

@vrutkovs why don't you sync with @deads2k or @soltysh on #1722 and ask them for their feedback ?

Reply:
#1722 fixes the thread level and https://github.com/openshift/cluster-kube-apiserver-operator/pull/1669 will fix process level races